README.md

```diff
@@ -1,138 +1,615 @@
+
+The `@logged` decorator logs a console message when a method starts and finishes. Many other popular decorators will also want to wrap a function, e.g., `@deprecated`, `@debounce`, `@memoize`, dependency injection, etc.
+
+Usage:
+
+```mjs
+import { @logged } from "./logged.mjs";
+
+class C {
+  @logged
+  method(arg) {
+    this.#x = arg;
   }
 
+  @logged
+  set #x(value) { }
+}
+
+new C().method(1);
+// starting method with arguments 1
+// starting set #x with arguments 1
+// ending set #x
+// ending method
+```
+
+`@logged` can be implemented in JavaScript in terms of built-in decorators: The `@logged` decorator is defined in terms of the `@wrap` built-in decorator, which takes a function as an argument. The method is passed through this function to get the method which is finally present on the class. `@wrap` is similar to core-decorators' [`@decorate`](https://www.npmjs.com/package/core-decorators#decorate) decorator.
+
+```mjs
+// logged.mjs
+
+export decorator @logged {
+  @wrap(f => {
+    const name = f.name;
+    function wrapped(...args) {
+      console.log(`starting ${name} with arguments ${args.join(", ")}`);
+      f.call(this, ...args);
+      console.log(`ending ${name}`);
+    }
+    wrapped.name = name;
+    return wrapped;
+  })
+}
+```
+
+In the above example, the *composed decorator* `@logged` is defined to expand out into a call of the `@wrap` decorator with a particular fixed callback.
+
+### `@defineElement`
+
+[HTML Custom Elements](https://developer.mozilla.org/en-US/docs/Web/Web_Components/Using_custom_elements) lets you define your own HTML element. Elements are registered using `customElements.define`. Using decorators, the registration can happen up-front:
+
+```mjs
+import { @defineElement } from "./defineElement.mjs";
+
+@defineElement('my-class')
+class MyClass extends HTMLElement { }
+```
+
+The `@defineElement` decorator is based on the `@register` decorator. This decorator is called when the class is finished being defined, and it calls the provided callback with the class that resulted.
+
+```mjs
+// defineElement.mjs
+export decorator @defineElement(name, options) {
+  @register(klass => customElements.define(name, klass, options))
+}
+```
+
+This example uses a *decorator arrow function* `@(args) => @decorators` which lets a decorator definition take arguments that can be used to supply arguments to other decorators in its definition.
+
+### `@metadata`
+
+The `@metdata(key, value)` decorator is similar to [`@Reflect.metadata`](https://github.com/rbuckton/reflect-metadata): It allows the easy retrival of information which was stored by annotating the class.
+
+```mjs
+import { @metadata } from "./metadata.mjs";
+
+// partially apply the decorator locally for terseness
+decorator @localMeta { @metadata("key", "value") }
+
+@localMeta class C {
+  @localMeta method() { }
+}
+
+Reflect.getMetadata(C, "key");                      // "value"
+Reflect.getMetadata(C.prototype, "key", "method");  // "value"
```

Comment:
it seems like this example shows a userland decorator, but one that has language-level support for "metadata". Should this instead import `getMetadata` from `./metadata.mjs`?

Reply:
There's a link to the Reflect.metadata proposal in the text.

Reply:
yes, but presumably that proposal would make this decorator built-in; since the example is showing how you'd implement it, it seems strange to assume `Reflect` has `getMetadata` on it.

Reply:
Not sure if you saw the `import "reflect-metadata";` statement in the decorator implementation. Anyway, I added a more specific note describing the use of the other proposal.